router/schema_validator/object_schema_validator.go

```diff
@@ -14,84 +14,93 @@ import (
 
 var textMarshallerType = reflect.TypeOf(new(encoding.TextMarshaler)).Elem()
 
-func (c typeSchemaValidatorContext) validateObjectSchema() error {
+func (c typeSchemaValidatorContext) validateObjectSchema() {
 	// TODO: validate required properties, nullable, additionalProperties, etc.
-	l := c.newLogger()
-	if c.schema.Type != objectSchemaType {
-		return nil
+	serializedFromObject := isSerializedFromObject(c.goType)
+
+	if !serializedFromObject {
+		if c.schema.Type == openapi3.TypeObject {
+			c.err(schemaTypeIsIncompatibleWithType(c.schema, c.goType))
+		}
+		return
+	}
+
+	if !isSchemaTypeObjectOrEmpty(c.schema) {
+		c.err(schemaTypeIsIncompatibleWithType(c.schema, c.goType))
 	}
+
+	handleMultiType(func(t reflect.Type) bool {
+		if t.Kind() == reflect.Struct {
+			return c.assertStruct(t)
+		}
+
+		// kind must be struct or object because we validated above serializedFromObject
+		return c.assertMap(t)
```

Comment:
c.assertMap(c.goType) will succeed, but not c.assertMap(t).
t can have kind bool

Reply:
Fixed